## Homework 2

### Due Friday, 9/20/2024

### For each homework set: 

### Create a New iPython (Jupyter) notebook. Name the notebook FirstAndLastName_Homework2 and save it before you start working

### To submit, export or print your notebook as a pdf, with all outputs visible. Upload both the pdf and a copy of your notebook (.ipynb) in Canvas.

1. Exercise 2.8: Create two arrays a and b<br>
    a = [1,2,3,4]<br>
    b = [2,4,6,8]<br>
    Compute the following:<br>
    
    b/a + 1<br>
    b/(a+1)<br>
    1/a<br>
    
    Verify these values by computing it by hand. (Either paper or using Markdown)

In [3]:
import numpy as np
import timeit

In [30]:
a = np.array([1, 2, 3, 4])
b = np.array([2, 4, 6, 8])
print(f"a = {a}")
print(f"b = {b}")
print(f"b/a + 1 = {b/a + 1}")
print(f"b/(a+1) = {b/(a+1)}")
print(f"1/a = {1/a}")

a = [1 2 3 4]
b = [2 4 6 8]
b/a + 1 = [3. 3. 3. 3.]
b/(a+1) = [1.         1.33333333 1.5        1.6       ]
1/a = [1.         0.5        0.33333333 0.25      ]


Solving for b/a + 1:<br>
b/a + 1 = [2/1 + 1, 4/2 + 1, 6/3 + 1, 8/4 + 1]<br>
b/a + 1 = [2 + 1, 2 + 1, 2 + 1, 2 + 1]<br>
b/a + 1 = [3, 3, 3, 3]<br>

Solving for b/(a+1):<br>
b/(a+1) = [2/(1+1), 4/(2+1), 6/(3+1), 8/(4+1)]<br>
b/(a+1) = [2/2, 4/3, 6/4, 8/5]<br>
b/(a+1) = [1, 1.3333, 1.5, 1.6]<br>

Solving for 1/a:<br>
1/a = [1/1, 1/2, 1/3, 1/4]<br>
1/a = [1, 0.5, 0.3333, 0.25]<br>

2. Using the same arrays above. Write a program to compute the dot product between them by accessing the elements of the array individually. Verify the result using the built-in function. (np.dot)   You may need to look up what a dot product is.

In [90]:
def dotProduct(arr1, arr2):
    product = 0
    for i in np.arange(arr1.size):
        print(f"arr1[{i}] = {arr1[i]}\narr2[{i}] = {arr2[i]}")
        print(f"{arr1[i]} * {arr2[i]} = {arr1[i] * arr2[i]}")
        print(f"{product} + {arr1[i] * arr2[i]} = {product + arr1[i] * arr2[i]}")
        product += arr1[i] * arr2[i]
        print(f"dot product so far is {product}\n")
    return product

dotProduct(a, b)
print(f"np.dot = {np.dot(a, b)}")

arr1[0] = 1
arr2[0] = 2
1 * 2 = 2
0 + 2 = 2
dot product so far is 2

arr1[1] = 2
arr2[1] = 4
2 * 4 = 8
2 + 8 = 10
dot product so far is 10

arr1[2] = 3
arr2[2] = 6
3 * 6 = 18
10 + 18 = 28
dot product so far is 28

arr1[3] = 4
arr2[3] = 8
4 * 8 = 32
28 + 32 = 60
dot product so far is 60

np.dot = 60


3. Download the Gaussian.txt found on Canvas and put it in the same directory as your code. <br>
    Load the values from the file using numpy.loadtxt()<br>
    Compute the following: the sum, length, mean of all the values in the file.<br>
    Then compute root mean square deviation (RMSD)<br>
    <p style="text-align: center;">$RMSD = \sqrt{\frac{\sum_{i=0}^{i=N}(mean-values[i])^2}{N}}$ </p><br>
    Since the values in the Gaussian.txt is drawn from a Gaussian distribution. The mean and RMSD should be almost the same as a Gaussian centered at 10 and with a sigma of 2.

In [10]:
x = np.loadtxt("Gaussian.txt")
sum = np.sum(x)
print(f"sum = {sum}")

length = x.size
print(f"length = {length}")

mean = np.mean(x)
print(f"mean = {mean}")

RMSD = np.sqrt(np.sum((x - mean)**2) / length)
print(f"RMSD = {RMSD}")

sum = 10037.52371
length = 1000
mean = 10.03752371
RMSD = 1.922871400919036


4. Download the matrix.txt found on Canvas and put in the same directory as your code <br>
    - Load the values from the file using numpy.loadtxt()
    - print out the following properties of the matrix and explain what each tell you about the array:
        - len, shape, sum, min, max (note that some of these function will require you loop through each row of the matrix)
    - Slice the matrix to print out only the 1st Row
    - Slice the matrix to print out only the 1st Column
    - Slice the matrix to print out only the 1st 3 Row
    - Slice the matrix to print out only the 1st 3 Column    
    - Slice the matrix to print out only the last 3 Row
    - Slice the matrix to print out only the last 3 Column

In [27]:
matrix = np.loadtxt("matrix.txt")
print(f"matrix = \n{matrix}\n")

print(f"matrix length = {matrix.size}\nLength describes how many elements are in the matrix.\n")
print(f"matrix shape = {matrix.shape}\nShape describes the dimensions of the array.\n")
print(f"matrix sum = {matrix.sum()}\nSum describes the value when all the elements in the matrix are added together.\n")
print(f"matrix min = {matrix.min()}\nMin describes the minimum value in the matrix.\n")
print(f"matrix max = {matrix.max()}\nMax describes the maximum value in the matrix.\n")
print(f"1st Row = {matrix[0, :]}\n")
print(f"1st Column = {matrix[:, 0]}\n")
print(f"First 3 Rows =\n{matrix[:3, :]}\n")
print(f"First 3 Columns =\n{matrix[:, :3]}\n")
print(f"Last 3 Rows =\n{matrix[-3:, :]}\n")
print(f"Last 3 Columns=\n{matrix[:, -3:]}\n")

matrix = 
[[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9.]
 [10. 11. 12. 13. 14. 15. 16. 17. 18. 19.]
 [20. 21. 22. 23. 24. 25. 26. 27. 28. 29.]
 [30. 31. 32. 33. 34. 35. 36. 37. 38. 39.]
 [40. 41. 42. 43. 44. 45. 46. 47. 48. 49.]
 [50. 51. 52. 53. 54. 55. 56. 57. 58. 59.]
 [60. 61. 62. 63. 64. 65. 66. 67. 68. 69.]
 [70. 71. 72. 73. 74. 75. 76. 77. 78. 79.]
 [80. 81. 82. 83. 84. 85. 86. 87. 88. 89.]
 [90. 91. 92. 93. 94. 95. 96. 97. 98. 99.]]

matrix length = 100
Length describes how many elements are in the matrix.

matrix shape = (10, 10)
Shape describes the dimensions of the array.

matrix sum = 4950.0
Sum describes the value when all the elements in the matrix are added together.

matrix min = 0.0
Min describes the minimum value in the matrix.

matrix max = 99.0
Max describes the maximum value in the matrix.

1st Row = [0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]

1st Column = [ 0. 10. 20. 30. 40. 50. 60. 70. 80. 90.]

First 3 Rows =
[[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9.]
 [10. 11. 12. 13. 14. 15

5. Exercise 2.9

In [11]:
L = 325
start_time = timeit.default_timer()

i, j, k = np.meshgrid(np.arange(-L, L+1), np.arange(-L, L+1), np.arange(-L, L+1), indexing='ij')

r = np.sqrt(i**2 + j**2 + k**2)
r[L, L, L] = 1

sign = np.where((i + j + k) % 2 == 0, 1, -1)

potentials = sign / r

total_sum = np.sum(potentials) - potentials[L, L, L]

end_time = timeit.default_timer()

elapsed_time = end_time - start_time

print(f"The Madelung constant for L = {L} is approximately {total_sum}")
print(f"Time = {elapsed_time}")

The Madelung constant for L = 325 is approximately -1.7493383281793022
Time = 25.134665900026448


6. Exercise 2.10

In [83]:
def calc_a5(A, Z):
    if (A % 2 == 1):
        return 0
    elif (A % 2 == 0 and Z % 2 == 0):
        return 12
    elif (A % 2 == 0 and Z % 2 == 1):
        return -12

def calc_b(A, Z, a5):
    a1 = 15.8
    a2 = 18.3
    a3 = 0.714
    a4 = 23.2
    return a1*A - a2*A**(2/3) - a3*(Z**2/A**(1/3)) - a4*((A - 2*Z)**2/A) + a5/A**(1/2)

def most_stable_A(Z):
    max_bepn = -1
    A_max = Z

    for A in range(Z, 3*Z + 1):
        a5 = calc_a5(A, Z)
        B = calc_b(A, Z, a5)
        bepn = B/A

        if bepn > max_bepn:
            max_bepn = bepn
            A_max = A

    return A_max, max_bepn

A = int(input("Enter Mass Number (A):"))
Z = int(input("Enter Atomic Number (Z):"))
a5 = calc_a5(A, Z)

B = calc_b(A, Z, a5)
print(f"Binding Energy (B) = {B}")

bepn = B/A
print(f"Binding Energy Per Nuclean = {bepn}")

Z = int(input("Enter Atomic Number (Z):"))

A_max, max_bepn = most_stable_A(Z)
print(f"The most stable nucleus has A = {A_max}")
print(f"Binding Energy Per Nucleon = {max_bepn}")

for Z in range(1, 101):
    A_max, max_bepn = most_stable_A(Z)    
    print(f"For Z = {Z}, the most stable A = {A_max}, with Binding Energy Per Nucleon = {max_bepn}")

Enter Mass Number (A): 58
Enter Atomic Number (Z): 28


Binding Energy (B) = 497.5620206224374
Binding Energy Per Nuclean = 8.578655527973059


Enter Atomic Number (Z): 28


The most stable nucleus has A = 62
Binding Energy Per Nucleon = 8.70245768367189
For Z = 1, the most stable A = 3, with Binding Energy Per Nucleon = 0.36869091831015827
For Z = 2, the most stable A = 4, with Binding Energy Per Nucleon = 5.321930578649441
For Z = 3, the most stable A = 7, with Binding Energy Per Nucleon = 5.280168164356119
For Z = 4, the most stable A = 8, with Binding Energy Per Nucleon = 6.466330085889912
For Z = 5, the most stable A = 11, with Binding Energy Per Nucleon = 6.650123444727665
For Z = 6, the most stable A = 14, with Binding Energy Per Nucleon = 7.200918138809924
For Z = 7, the most stable A = 15, with Binding Energy Per Nucleon = 7.330860591990981
For Z = 8, the most stable A = 18, with Binding Energy Per Nucleon = 7.719275577459026
For Z = 9, the most stable A = 19, with Binding Energy Per Nucleon = 7.73697768275634
For Z = 10, the most stable A = 22, with Binding Energy Per Nucleon = 8.035350864715019
For Z = 11, the most stable A = 25, with Binding En